src/enable_variable_ops/ngraph_rewrite_pass.cc

```diff
@@ -0,0 +1,330 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+#include <iomanip>
+
+#include "tensorflow/core/common_runtime/optimization_registry.h"
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/graph/graph.h"
+
+#include "ngraph_api.h"
+#include "ngraph_assign_clusters.h"
+#include "ngraph_capture_variables.h"
+#include "ngraph_deassign_clusters.h"
+#include "ngraph_encapsulate_clusters.h"
+#include "ngraph_enter_in_catalog.h"
+#include "ngraph_log.h"
+#include "ngraph_mark_for_clustering.h"
+#include "ngraph_replace_optimizers.h"
+#include "ngraph_rewrite_for_tracking.h"
+#include "tf_graph_writer.h"
+
+#if defined NGRAPH_DISTRIBUTED
+#include "ngraph/distributed.hpp"
+#endif
+
+using namespace std;
+
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
+class NGraphRewritePass : public GraphOptimizationPass {
+ public:
+  virtual Status Run(const GraphOptimizationPassOptions& options) = 0;
+
+ protected:
+  void DumpGraphs(const GraphOptimizationPassOptions& options, int idx,
+                  std::string filename_prefix, std::string title) {
+    // If we have a "main" graph, dump that.
+    if (options.graph != nullptr) {
+      auto dot_filename = DotFilename(filename_prefix, idx);
+      auto pbtxt_filename = PbtxtFilename(filename_prefix, idx);
+      NGRAPH_VLOG(0) << "Dumping main graph to " << dot_filename;
+      NGRAPH_VLOG(0) << "Dumping main graph to " << pbtxt_filename;
+
+      GraphToDotFile(options.graph->get(), dot_filename, title);
+      GraphToPbTextFile(options.graph->get(), pbtxt_filename);
+    }
+
+    // If we have partition graphs (we shouldn't), dump those.
+    if (options.partition_graphs != nullptr) {
+      int sub_idx = 0;
+
+      for (auto& kv : *options.partition_graphs) {
+        auto dot_filename = DotFilename(filename_prefix, idx, sub_idx);
+        auto pbtxt_filename = PbtxtFilename(filename_prefix, idx, sub_idx);
+        NGRAPH_VLOG(0) << "Dumping subgraph " << sub_idx << " to "
+                       << dot_filename;
+        NGRAPH_VLOG(0) << "Dumping subgraph " << sub_idx << " to "
+                       << pbtxt_filename;
+
+        Graph* pg = kv.second.get();
+
+        GraphToDotFile(pg, dot_filename, title);
+        GraphToPbTextFile(pg, pbtxt_filename);
+
+        sub_idx++;
+      }
+    }
+  }
+
+  // Returns a fresh "serial number" to avoid filename collisions in the graph
+  // dumps.
+  static int FreshIndex() {
+    mutex_lock l(s_serial_counter_mutex);
+    return s_serial_counter++;
+  }
+
+  static bool DumpAllGraphs() {
```

Comment:
These functions are defined in ngraph_utils so if you include that you will not need these here.